packager/packager.cc

```diff
@@ -231,6 +232,95 @@ ast::UnresolvedConstantLit *verifyConstant(core::MutableContext ctx, core::NameR
     return target;
 }
 
+struct EnforcePackagePrefix {
+    const PackageInfo *pkg;
+    vector<core::NameRef> nameParts;
+    int rootConsts = 0;
+
+    EnforcePackagePrefix(const PackageInfo *pkg) : pkg(pkg) {
+        ENFORCE(pkg != nullptr);
+    }
+
+    void pushConstantLit(ast::UnresolvedConstantLit *lit) {
+        auto oldLen = nameParts.size();
+        while (lit != nullptr) {
+            nameParts.emplace_back(lit->cnst);
+            auto scope = ast::cast_tree<ast::ConstantLit>(lit->scope);
+            lit = ast::cast_tree<ast::UnresolvedConstantLit>(lit->scope);
+            if (scope != nullptr) {
+                ENFORCE(lit == nullptr);
+                ENFORCE(scope->symbol == core::Symbols::root());
+                rootConsts++;
+            }
+        }
+        reverse(nameParts.begin() + oldLen, nameParts.end()); // TODO I could do this recursively
+    }
+
+    void popConstantLit(ast::UnresolvedConstantLit *lit) {
+        while (lit != nullptr) {
+            nameParts.pop_back();
+            auto scope = ast::cast_tree<ast::ConstantLit>(lit->scope);
+            lit = ast::cast_tree<ast::UnresolvedConstantLit>(lit->scope);
+            if (scope != nullptr) {
+                ENFORCE(lit == nullptr);
+                ENFORCE(scope->symbol == core::Symbols::root());
+                rootConsts--;
+            }
+        }
+    }
+
+    ast::ExpressionPtr preTransformClassDef(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto &classDef = ast::cast_tree_nonnull<ast::ClassDef>(tree);
+        if (classDef.symbol == core::Symbols::root()) {
+            // Ignore top-level <root>
+            return tree;
+        }
+        auto &pkgName = pkg->name.fullName.parts;
+        bool skipCheck = nameParts.size() > pkgName.size();
+        auto *constantLit = &ast::cast_tree_nonnull<ast::UnresolvedConstantLit>(classDef.name);
```

Comment:
You probably just want to take a reference here:

```suggestion
        auto &constantLit = ast::cast_tree_nonnull<ast::UnresolvedConstantLit>(classDef.name);
```